File: leetcode/904.水果成篮/904.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        i, res,substring = 0, 0, 0
        cnt = defaultdict(int)
        for j, c in enumerate(fruits):
            cnt[c] += 1
            while len(cnt) > 2:
                cnt[fruits[i]] -= 1
                if(cnt[fruits[i]] == 0):
                    del cnt[fruits[i]]
                i += 1
            substring = j - i + 1
            res = substring if res < substring else res
        return res 
        

File: leetcode/904.水果成篮/904.java
class Solution {
    public int totalFruit(int[] fruits) {
        int i = 0;
        int substring = 0;
        int res = 0;
        Map<Integer, Integer> cnt = new HashMap<>();
        for(int j = 0; j < fruits.length; j++){
            cnt.merge(fruits[j], 1, Integer::sum);
            while(cnt.size() > 2){
                cnt.merge(fruits[i], -1, Integer::sum);
                if(cnt.get(fruits[i]) == 0){
                    cnt.remove(fruits[i]);
                }
                i++;
            }
            substring = j-i+1;
            res = res <substring ? substring: res;
        }
        return res;
    }
}